File: JavaSimpleBackend/src/main/java/com/atesti/service/StatisticsService.java
package com.atesti.service;

import com.atesti.dto.*;
import com.atesti.entity.Aktivnost;
import com.atesti.entity.RadniNalog;
import com.atesti.entity.User;
import com.atesti.repository.*;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class StatisticsService {

    private final RadniNalogRepository radniNalogRepository;
    private final NaruciteljRepository naruciteljRepository;
    private final AktivnostRepository aktivnostRepository;
    private final UserRepository userRepository;
    private final ObjectMapper objectMapper;

    public DashboardStatsResponse getDashboardStats() {
        long totalNalozi = radniNalogRepository.count();
        long totalNarucitelji = naruciteljRepository.count();
        long fakturirano = radniNalogRepository.countByFakturirano(true);
        long zavrseno = radniNalogRepository.countByZavrseno(true);

        return DashboardStatsResponse.builder()
                .totalNalozi(totalNalozi)
                .totalNarucitelji(totalNarucitelji)
                .fakturirano(fakturirano)
                .zavrseno(zavrseno)
                .nefakturirano(totalNalozi - fakturirano)
                .uTijeku(totalNalozi - zavrseno)
                .build();
    }

    public List<RevenueItemResponse> getRevenueByAktivnost() {
        List<Aktivnost> aktivnosti = aktivnostRepository.findByIsActiveTrueOrderByIdAsc();
        List<RadniNalog> nalozi = radniNalogRepository.findAll();

        Map<Long, Long> countMap = new HashMap<>();
        for (RadniNalog nalog : nalozi) {
            List<Long> ids = parseAktivnostiJson(nalog.getAktivnosti());
            for (Long id : ids) {
                countMap.merge(id, 1L, Long::sum);
            }
        }

        return aktivnosti.stream().map(a -> {
            long count = countMap.getOrDefault(a.getId(), 0L);
            double cijena = a.getCijena() != null ? a.getCijena().doubleValue() : 0;
            return RevenueItemResponse.builder()
                    .id(a.getId())
                    .aktivnost(a.getAktivnost())
                    .cijena(cijena)
                    .count(count)
                    .ukupno(cijena * count)
                    .build();
        }).collect(Collectors.toList());
    }

    public List<PerformanceItemResponse> getPerformanceByWorker() {
        List<User> users = userRepository.findAllByOrderByIdAsc();
        List<RadniNalog> assigned = radniNalogRepository.findByAssignedUserIdIsNotNull();

        Map<Long, List<RadniNalog>> byUser = assigned.stream()
                .collect(Collectors.groupingBy(n -> n.getAssignedUser().getId()));

        return users.stream().map(u -> {
            List<RadniNalog> userNalozi = byUser.getOrDefault(u.getId(), List.of());
            return PerformanceItemResponse.builder()
                    .id(u.getId())
                    .username(u.getUsername())
                    .ime(u.getIme())
                    .prezime(u.getPrezime())
                    .total(userNalozi.size())
                    .zavrseno(userNalozi.stream().filter(n -> Boolean.TRUE.equals(n.getZavrseno())).count())
                    .fakturirano(userNalozi.stream().filter(n -> Boolean.TRUE.equals(n.getFakturirano())).count())
                    .build();
        }).collect(Collectors.toList());
    }

    public List<MonthlyItemResponse> getIssuedByMonth() {
        List<RadniNalog> nalozi = radniNalogRepository.findAll();
        DateTimeFormatter fmt = DateTimeFormatter.ofPattern("yyyy-MM");

        Map<String, Long> grouped = nalozi.stream()
                .filter(n -> n.getDatum() != null)
                .collect(Collectors.groupingBy(
                        n -> n.getDatum().format(fmt),
                        TreeMap::new,
                        Collectors.counting()));

        return grouped.entrySet().stream()
                .map(e -> MonthlyItemResponse.builder()
                        .month(e.getKey())
                        .count(e.getValue())
                        .build())
                .collect(Collectors.toList());
    }

    private List<Long> parseAktivnostiJson(String json) {
        if (json == null || json.isBlank()) return List.of();
        try {
            return objectMapper.readValue(json, new TypeReference<List<Long>>() {});
        } catch (Exception e) {
            return List.of();
        }
    }
}

File: backend/services/statisticsService.js
const { Sequelize } = require('sequelize');
const db = require('../models');
const RadniNalog = db.radni_nalog;
const Narucitelj = db.narucitelj;
const Aktivnost = db.aktivnost;
const User = db.user;

const getDashboardStats = async () => {
  const totalNalozi = await RadniNalog.count();
  const totalNarucitelji = await Narucitelj.count();
  const fakturirano = await RadniNalog.count({ where: { fakturirano: true } });
  const zavrseno = await RadniNalog.count({ where: { zavrseno: true } });
  const nefakturirano = await RadniNalog.count({ where: { fakturirano: false } });
  const uTijeku = await RadniNalog.count({ where: { zavrseno: false } });

  return { totalNalozi, totalNarucitelji, fakturirano, zavrseno, nefakturirano, uTijeku };
};

const getRevenueByAktivnost = async () => {
  const aktivnosti = await Aktivnost.findAll({
    where: { isActive: true },
    attributes: ['id', 'aktivnost', 'cijena'],
    order: [['aktivnost', 'ASC']]
  });

  const nalozi = await RadniNalog.findAll({ attributes: ['aktivnosti'] });

  // Count by activity ID
  const countById = {};
  for (const nalog of nalozi) {
    let aktivnostiArr = [];
    if (typeof nalog.aktivnosti === 'string') {
      try {
        aktivnostiArr = JSON.parse(nalog.aktivnosti);
      } catch {
        aktivnostiArr = [];
      }
    } else if (Array.isArray(nalog.aktivnosti)) {
      aktivnostiArr = nalog.aktivnosti;
    }
    for (const id of aktivnostiArr) {
      if (id) countById[id] = (countById[id] || 0) + 1;
    }
  }

  return aktivnosti.map(a => {
    const count = countById[a.id] || 0;
    const cijena = a.cijena ? parseFloat(a.cijena) : 0;
    return {
      id: a.id,
      aktivnost: a.aktivnost,
      cijena,
      count,
      ukupno: cijena * count
    };
  });
};

const getPerformanceByWorker = async () => {
  const users = await User.findAll({
    attributes: { exclude: ['password'] }
  });

  const nalozi = await RadniNalog.findAll({
    where: { assignedUserId: { [Sequelize.Op.ne]: null } },
    attributes: ['assignedUserId', 'zavrseno', 'fakturirano']
  });

  const statsById = {};
  for (const nalog of nalozi) {
    const uid = nalog.assignedUserId;
    if (!statsById[uid]) statsById[uid] = { total: 0, zavrseno: 0, fakturirano: 0 };
    statsById[uid].total++;
    if (nalog.zavrseno) statsById[uid].zavrseno++;
    if (nalog.fakturirano) statsById[uid].fakturirano++;
  }

  return users.map(u => ({
    id: u.id,
    username: u.username,
    ime: u.ime,
    prezime: u.prezime,
    total: statsById[u.id]?.total || 0,
    zavrseno: statsById[u.id]?.zavrseno || 0,
    fakturirano: statsById[u.id]?.fakturirano || 0
  }));
};

const getIssuedByMonth = async () => {
  const nalozi = await RadniNalog.findAll({
    attributes: ['datum'],
    order: [['datum', 'ASC']]
  });

  const byMonth = {};
  for (const nalog of nalozi) {
    if (!nalog.datum) continue;
    const d = new Date(nalog.datum);
    const key = `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}`;
    byMonth[key] = (byMonth[key] || 0) + 1;
  }

  return Object.entries(byMonth).map(([month, count]) => ({ month, count }));
};

module.exports = { getDashboardStats, getRevenueByAktivnost, getPerformanceByWorker, getIssuedByMonth };
